frontend/src/metabase/lib/expressions/resolver.js

```diff
@@ -85,7 +91,8 @@ export function resolve(expression, type = "expression", fn = undefined) {
     } else if (op === "true" || op === "false") {
       operandType = "expression";
     } else if (COMPARISON_OPS.includes(op)) {
-      operandType = "expression";
+      const isEquality = op === OP.Equal || op === OP.NotEqual;
+      operandType = isEquality ? "expression" : "value";
```

Comment:
For non equality, the comparable type is a new one called `value` (which is compatible only with `number` and `string`).